pkg/watchdog/daemon.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package watchdog
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+const daemonBin = "watchdogd"
+
+// DaemonOpts contain options for the watchdog daemon.
+type DaemonOpts struct {
+	// Dev is the watchdog device. Ex: /dev/watchdog
+	Dev string
+
+	// When set to -1, uses the preset values.
+	TimeoutSecs, PreTimeoutSecs int
+
+	// KeepAliveSecs is the length of the keep alive interval.
+	KeepAliveSecs int
+}
+
+// Run runs the watchdog on the current goroutine.
+//
+// It starts armed, and can be controlled with signals:
+// - USR1: disarm
+// - USR2: arm
+//
+// Cancelling the context will exit with the current armed/disarmed state.
+func Run(ctx context.Context, opts *DaemonOpts) error {
```

Comment:
Can we have a mechanism to make sure this is only called once within a process? Maybe a global variable mutex or something that'd make Run return "watchdog petting already running", since SIGUSR1 and SIGUSR2 are only a once per process thing.

Reply:
Is this necessary? The docs say "It is allowed to call Notify multiple times with different channels and the same signals: each channel receives copies of incoming signals independently." You can do things like running a daemon for /dev/watchdog0 and /dev/watchdog1 in the same process and disarm both at once.